File: templates/Wpf/_comp/_shared/Service.IdentityCommon/Param_ProjectName.Core/Services/IdentityService.cs
using System;
using System.IO;
using System.Linq;
using System.Net.NetworkInformation;
using System.Reflection;
using System.Security.Cryptography;
using System.Threading.Tasks;
using Microsoft.Identity.Client;
using Param_RootNamespace.Core.Contracts.Services;
using Param_RootNamespace.Core.Helpers;

namespace Param_RootNamespace.Core.Services
{
    public class IdentityService : IIdentityService
    {
        /*
        For more information about using Identity, see
        https://github.com/microsoft/WindowsTemplateStudio/blob/release/docs/WPF/services/identity.md

        Read more about Microsoft Identity Client here
        https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki
        https://docs.microsoft.com/azure/active-directory/develop/v2-overview

        TODO WTS: Please create a ClientID following these steps and update the appsettings.json IdentityClientId.
        https://docs.microsoft.com/azure/active-directory/develop/quickstart-register-app

        The provided clientID requests permissions on user.read, this might be blocked in environments that require admin consent.
        For more info about admin consent please see https://docs.microsoft.com/azure/active-directory/develop/application-consent-experience
        For more info creating protected APIs, see https://docs.microsoft.com/azure/active-directory/develop/scenario-protected-web-api-overview
        For more info on desktop apps that call protected APIs, see https://docs.microsoft.com/azure/active-directory/develop/scenario-desktop-overview
        */
        private readonly string[] _graphScopes = new string[] { "user.read" };
        private readonly IIdentityCacheService _identityCacheService;
        private bool _integratedAuthAvailable;
        private IPublicClientApplication _client;
        private AuthenticationResult _authenticationResult;

        public event EventHandler LoggedIn;

        public event EventHandler LoggedOut;

        public IdentityService(IIdentityCacheService identityCacheService)
        {
            _identityCacheService = identityCacheService;
        }

        public void InitializeWithAadAndPersonalMsAccounts(string clientId, string redirectUri = null)
        {
            _integratedAuthAvailable = false;
            _client = PublicClientApplicationBuilder.Create(clientId)
                                                    .WithAuthority(AadAuthorityAudience.AzureAdAndPersonalMicrosoftAccount)
                                                    .WithRedirectUri(redirectUri)
                                                    .Build();

            ConfigureCache();
        }

        public void InitializeWithAadMultipleOrgs(string clientId, bool integratedAuth = false, string redirectUri = null)
        {
            _integratedAuthAvailable = integratedAuth;
            _client = PublicClientApplicationBuilder.Create(clientId)
                                                    .WithAuthority(AadAuthorityAudience.AzureAdMultipleOrgs)
                                                    .WithRedirectUri(redirectUri)
                                                    .Build();

            ConfigureCache();
        }

        public void InitializeWithAadSingleOrg(string clientId, string tenant, bool integratedAuth = false, string redirectUri = null)
        {
            _integratedAuthAvailable = integratedAuth;
            _client = PublicClientApplicationBuilder.Create(clientId)
                                                    .WithAuthority(AzureCloudInstance.AzurePublic, tenant)
                                                    .WithRedirectUri(redirectUri)
                                                    .Build();

            ConfigureCache();
        }

        public bool IsLoggedIn() => _authenticationResult != null;

        public async Task<LoginResultType> LoginAsync()
        {
            if (!NetworkInterface.GetIsNetworkAvailable())
            {
                return LoginResultType.NoNetworkAvailable;
            }

            try
            {
                var accounts = await _client.GetAccountsAsync();
                _authenticationResult = await _client.AcquireTokenInteractive(_graphScopes)
                                                     .WithAccount(accounts.FirstOrDefault())
                                                     .ExecuteAsync();
                if (!IsAuthorized())
                {
                    _authenticationResult = null;
                    return LoginResultType.Unauthorized;
                }

                LoggedIn?.Invoke(this, EventArgs.Empty);
                return LoginResultType.Success;
            }
            catch (MsalClientException ex)
            {
                if (ex.ErrorCode == "authentication_canceled")
                {
                    return LoginResultType.CancelledByUser;
                }

                return LoginResultType.UnknownError;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine(ex);
                return LoginResultType.UnknownError;
            }
        }

        public bool IsAuthorized()
        {
            // TODO WTS: You can also add extra authorization checks here.
            // i.e.: Checks permisions of _authenticationResult.Account.Username in a database.
            return true;
        }

        public string GetAccountUserName()
        {
            return _authenticationResult?.Account?.Username;
        }

        public async Task LogoutAsync()
        {
            try
            {
                var accounts = await _client.GetAccountsAsync();
                var account = accounts.FirstOrDefault();
                if (account != null)
                {
                    await _client.RemoveAsync(account);
                }

                _authenticationResult = null;
                LoggedOut?.Invoke(this, EventArgs.Empty);
            }
            catch (MsalException)
            {
                // TODO WTS: LogoutAsync can fail please handle exceptions as appropriate to your scenario
                // For more info on MsalExceptions see
                // https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/exceptions
            }
        }

        public async Task<string> GetAccessTokenForGraphAsync() => await GetAccessTokenAsync(_graphScopes);

        // All sensitive data in your app should be retrieven using access tokens.
        // This method provides you with an access token to secure calls to the Microsoft Graph or other protected API.
        // For more info on protecting web api with tokens see https://docs.microsoft.com/azure/active-directory/develop/scenario-protected-web-api-overview
        public async Task<string> GetAccessTokenAsync(string[] scopes)
        {
            var acquireTokenSuccess = await AcquireTokenSilentAsync(scopes);
            if (acquireTokenSuccess)
            {
                return _authenticationResult.AccessToken;
            }
            else
            {
                try
                {
                    // Interactive authentication is required
                    var accounts = await _client.GetAccountsAsync();
                    _authenticationResult = await _client.AcquireTokenInteractive(scopes)
                                                         .WithAccount(accounts.FirstOrDefault())
                                                         .ExecuteAsync();
                    return _authenticationResult.AccessToken;
                }
                catch (MsalException)
                {
                    // AcquireTokenSilent and AcquireTokenInteractive failed, the session will be closed.
                    _authenticationResult = null;
                    LoggedOut?.Invoke(this, EventArgs.Empty);
                    return string.Empty;
                }
            }
        }

        public async Task<bool> AcquireTokenSilentAsync() => await AcquireTokenSilentAsync(_graphScopes);

        private async Task<bool> AcquireTokenSilentAsync(string[] scopes)
        {
            if (!NetworkInterface.GetIsNetworkAvailable())
            {
                return false;
            }

            try
            {
                var accounts = await _client.GetAccountsAsync();
                _authenticationResult = await _client.AcquireTokenSilent(scopes, accounts.FirstOrDefault())
                                                     .ExecuteAsync();
                return true;
            }
            catch (MsalUiRequiredException ex)
            {
                System.Diagnostics.Debug.WriteLine(ex);
                if (_integratedAuthAvailable)
                {
                    try
                    {
                        _authenticationResult = await _client.AcquireTokenByIntegratedWindowsAuth(_graphScopes)
                                                             .ExecuteAsync();
                        return true;
                    }
                    catch (MsalUiRequiredException)
                    {
                        // Interactive authentication is required
                        return false;
                    }
                }
                else
                {
                    // Interactive authentication is required
                    return false;
                }
            }
            catch (MsalException)
            {
                // TODO WTS: Silentauth failed, please handle this exception as appropriate to your scenario
                // For more info on MsalExceptions see
                // https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/exceptions
                return false;
            }
        }

        private void ConfigureCache()
        {
            if (_identityCacheService != null)
            {
                // .NET core applications should provide a mechanism to serialize and deserialize the user token cache
                // https://aka.ms/msal-net-token-cache-serialization
                _client.UserTokenCache.SetBeforeAccess((args) =>
                {
                    var data = _identityCacheService.ReadMsalToken();
                    if (data != default)
                    {
                        args.TokenCache.DeserializeMsalV3(data);
                    }
                });
                _client.UserTokenCache.SetAfterAccess((args) =>
                {
                    if (args.HasStateChanged)
                    {
                        _identityCacheService.SaveMsalToken(args.TokenCache.SerializeMsalV3());
                    }
                });
            }
        }
    }
}

File: templates/Uwp/_comp/_shared/Service.IdentityForcedLogin._VB/Param_ProjectName.Core/Services/IdentityService.vb
Imports System.Configuration
Imports System.Net.NetworkInformation
Imports Param_RootNamespace.Core.Helpers
Imports Microsoft.Identity.Client

Namespace Services
    Public Class IdentityService

        ' For more information about using Identity, see
        ' https://github.com/Microsoft/WindowsTemplateStudio/blob/release/docs/UWP/services/identity.md
        '
        ' Read more about Microsoft Identity Client here
        ' https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki
        ' https://docs.microsoft.com/azure/active-directory/develop/v2-overview
        Private ReadOnly _graphScopes As String() = {"user.read"}
        Private _integratedAuthAvailable As Boolean
        Private _client As IPublicClientApplication
        Private _authenticationResult As AuthenticationResult

        ' TODO WTS: Please create a ClientID following these steps and update the app.config IdentityClientId.
        ' https://docs.microsoft.com/azure/active-directory/develop/quickstart-register-app
        ' Make sure you configure urn:ietf:wg:oauth:2.0:oob as a redirect uri as described in
        ' https://docs.microsoft.com/en-us/dotnet/api/microsoft.identity.client.applicationoptions.redirecturi?view=azure-dotnet
        Private _clientId As String = ConfigurationManager.AppSettings("IdentityClientId")

        Public Event LoggedIn As EventHandler

        Public Event LoggedOut As EventHandler

        Public Sub InitializeWithAadAndPersonalMsAccounts()
            _integratedAuthAvailable = False
            _client = PublicClientApplicationBuilder.Create(_clientId).WithAuthority(AadAuthorityAudience.AzureAdAndPersonalMicrosoftAccount).Build()
        End Sub

        Public Sub InitializeWithAadMultipleOrgs(Optional integratedAuth As Boolean = False)
            _integratedAuthAvailable = integratedAuth
            _client = PublicClientApplicationBuilder.Create(_clientId).WithAuthority(AadAuthorityAudience.AzureAdMultipleOrgs).Build()
        End Sub

        Public Sub InitializeWithAadSingleOrg(tenant As String, Optional integratedAuth As Boolean = False)
            _integratedAuthAvailable = integratedAuth
            _client = PublicClientApplicationBuilder.Create(_clientId).WithAuthority(AzureCloudInstance.AzurePublic, tenant).Build()
        End Sub

        Public Function IsLoggedIn() As Boolean
            Return _authenticationResult IsNot Nothing
        End Function

        Public Async Function LoginAsync() As Task(Of LoginResultType)
            If Not NetworkInterface.GetIsNetworkAvailable() Then
                Return LoginResultType.NoNetworkAvailable
            End If

            Try
                Dim accounts = Await _client.GetAccountsAsync()
                _authenticationResult = Await _client.AcquireTokenInteractive(_graphScopes).WithAccount(accounts.FirstOrDefault()).ExecuteAsync()

                If Not IsAuthorized() Then
                    _authenticationResult = Nothing
                    Return LoginResultType.Unauthorized
                End If

                RaiseEvent LoggedIn(Me, EventArgs.Empty)
                Return LoginResultType.Success
            Catch ex As MsalClientException

                If ex.ErrorCode = "authentication_canceled" Then
                    Return LoginResultType.CancelledByUser
                End If

                Return LoginResultType.UnknownError
            Catch __unusedException2__ As Exception
                Return LoginResultType.UnknownError
            End Try
        End Function

        Public Function IsAuthorized() As Boolean
            ' TODO WTS: You can also add extra authorization checks here.
            ' i.e.: Checks permisions of _authenticationResult.Account.Username in a database.
            Return True
        End Function

        Public Function GetAccountUserName() As String
            Return _authenticationResult?.Account?.Username
        End Function

        Public Async Function LogoutAsync() As Task
            Try
                Dim accounts = Await _client.GetAccountsAsync()
                Dim account = accounts.FirstOrDefault()

                If account IsNot Nothing Then
                    Await _client.RemoveAsync(account)
                End If

                _authenticationResult = Nothing
                RaiseEvent LoggedOut(Me, EventArgs.Empty)
            Catch msalException As MsalException
                ' TODO WTS: LogoutAsync can fail please handle exceptions as appropriate to your scenario
                ' For more info on MsalExceptions see
                ' https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/exceptions
            End Try
        End Function

        Public Async Function GetAccessTokenAsync(scopes As String()) As Task(Of String)
            Dim acquireTokenSuccess = Await AcquireTokenSilentAsync(scopes)

            If acquireTokenSuccess Then
                Return _authenticationResult.AccessToken
            Else
                Try
                    ' Interactive authentication is required
                    Dim accounts = Await _client.GetAccountsAsync()
                    _authenticationResult = Await _client.AcquireTokenInteractive(scopes).WithAccount(accounts.FirstOrDefault()).ExecuteAsync()
                    Return _authenticationResult.AccessToken
                Catch msalException As MsalException
                    ' AcquireTokenSilent and AcquireTokenInteractive failed, the session will be closed.
                    _authenticationResult = Nothing
                    RaiseEvent LoggedOut(Me, EventArgs.Empty)
                    Return String.Empty
                End Try
            End If
        End Function

        Public Async Function GetAccessTokenForGraphAsync() As Task(Of String)
            Return Await GetAccessTokenAsync(_graphScopes)
        End Function

        Public Async Function AcquireTokenSilentAsync() As Task(Of Boolean)
            Return Await AcquireTokenSilentAsync(_graphScopes)
        End Function

        Private Async Function AcquireTokenSilentAsync(scopes As String()) As Task(Of Boolean)
            If Not NetworkInterface.GetIsNetworkAvailable() Then
                Return False
            End If

            Dim retryWithUI As Boolean = False

            Try
                Dim accounts = Await _client.GetAccountsAsync()
                _authenticationResult = Await _client.AcquireTokenSilent(scopes, accounts.FirstOrDefault()).ExecuteAsync()
                Return True
            Catch ex As MsalUiRequiredException

                If _integratedAuthAvailable Then
                    retryWithUI = True
                Else
                    ' Interactive authentication is required
                    Return False
                End If

            Catch ex As MsalException
                ' TODO WTS: Silentauth failed, please handle this exception as appropriate to your scenario
                ' For more info on MsalExceptions see
                ' https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/exceptions
                Return False
            End Try

            If retryWithUI Then
                Try
                    _authenticationResult = Await _client.AcquireTokenByIntegratedWindowsAuth(scopes).ExecuteAsync()
                    Return True
                Catch ex As MsalUiRequiredException
                    ' Interactive authentication is required
                    Return False
                End Try
            Else
                Return False
            End If
        End Function
    End Class
End Namespace
